azure-iot-device/azure/iot/device/common/pipeline/pipeline_thread.py

```diff
@@ -0,0 +1,196 @@
+  pipeline thread is only entered in places where we know that external code is
+  calling into the pipeline (such as a client API call or a callback from a
+  third-party library).  Likewise, the callback thread is only entered in places
+  where we know that the pipeline is calling back into client code.
+
+5. Exceptions raised from the pipeline thread are still able to be caught by
+  the function which entered the pipeline thread.
+
+5. Calls into the pipeline thread can either block or not block.  Blocking is used
+  for cases where the caller needs a return value from the pipeline or is
+  expecting to handle any errors raised from the pipeline thread.  Blocking is
+  not used when the code calling into the pipeline is not waiting for a response
+  and is not expecting to handle any exceptions, such as protocol library
+  handlers which call into the pipeline to deliver protocol messages.
+
+6. Calls into the callback thread could theoretically block, but we currently
+  only have decorators which enter the callback thread without blocking.  This
+  is done to ensure that client code does not execute on the pipeline thread and
+  also to ensure that the pipline thread is not blocked while waiting for client
+  code to execute.
+
+These decorators use concurrent.futures.Future and the ThreadPoolExecutor because:
+
+1. The thread pooling with a pool size of 1 gives us a single thread to run all
+  pipeline operations and a different (single) thread to run all callbacks.  If
+  the code attempts to run a second pipeline operation (or callback) while a
+  different one is running, the ThreadPoolExecutor will queue the code until the
+  first call is completed.
+
+2. The concurent.futures.Future object properly handles both Exception and
+  BaseException errors, re-raising them when the Future.result method is called.
+  threading.Thread.get() was not an option because it doesn't re-raise
+  BaseException errors when Thread.get is called.
+
+3. concurrent.futures is available as a backport to 2.7.
+
+"""
+
+_executors = {}
+
+
+def _get_named_executor(thread_name):
+    """
+    Get a ThreadPoolExecutor object with the given name.  If no such executor exists,
+    this function will create on with a single worker and assign it to the provided
+    name.
+    """
+    global _executors
+    if thread_name not in _executors:
+        logger.info("Creating {} executor".format(thread_name))
+        _executors[thread_name] = ThreadPoolExecutor(max_workers=1)
+    return _executors[thread_name]
+
+
+def _invoke_on_executor_thread(func, thread_name, block=True):
+    """
+    Return wrapper to run the function on a given thread.  If block==False,
+    the call returns immediately without waiting for the decorated function to complete.
+    If block==True, the call waits for the decorated function to complete before returning.
+    """
+
+    # Mocks on py27 don't have a __name__ attribute.  Use str() if you can't use __name__
```

Comment:
>Mocks on py27 don't have a __name__ attribute.  Use str() if you can't use __name__ [](start = 6, length = 83)

I'm not sure if philosophically, test-related compat issues should be handled in the code under test. Couldn't the testing module manually attach a `__name__` attribute in 2.7?

Reply:
I completely agree.  I tried this and other parts of the test broke.  I don't remember the details, but it appeared as if someone was using the __name__ attribute to distinguish a mock from a real object.  My ability to experiment was somewhat hampered  by the fact that __name__ is special.  For example, I could set it, but I could not clear it -- the interpreter gave me an error if I tried to del it.  This told me that I was perhaps playing with fire and shouldn't.  

---
In reply to: [305460661](https://github.com/Azure/azure-iot-sdk-python-preview/pull/133#discussion_r305460661) [](ancestors = 305460661)